api/src/org/labkey/api/reader/TabBufferedReader.java

```
@@ -0,0 +1,157 @@
+
+    private TabBufferedReader(@NotNull Reader in, int initialBufferSize, int maxBufferSize)
+    {
+        super(in, initialBufferSize);
+        _currentBufferSize = initialBufferSize;
+        _maxBufferSize = maxBufferSize;
+
+        try
+        {
+            _readAheadLimitField = BufferedReader.class.getDeclaredField("readAheadLimit");
+            _readAheadLimitField.setAccessible(true);
+            _nextCharField = BufferedReader.class.getDeclaredField("nextChar");
+            _nextCharField.setAccessible(true);
+            _markedCharField = BufferedReader.class.getDeclaredField("markedChar");
+            _markedCharField.setAccessible(true);
+        }
+        catch (NoSuchFieldException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public TabBufferedReader(@NotNull Reader in, long exactCharacterCount)
+    {
+        this(in, getInitialBufferSize(exactCharacterCount), getMaxBufferSize(exactCharacterCount));
+    }
+
+    public TabBufferedReader(@NotNull Reader in)
+    {
+        this(in, INITIAL_BUFFER_SIZE, ABSOLUTE_MAX_BUFFER_SIZE);
+    }
+
+    private static int getInitialBufferSize(long exactCharacterCount)
+    {
+        return (int)Math.min(exactCharacterCount, INITIAL_BUFFER_SIZE);
+    }
+
+    private static int getMaxBufferSize(long exactCharacterCount)
+    {
+        return (int)Math.min(exactCharacterCount, ABSOLUTE_MAX_BUFFER_SIZE);
+    }
+
+    public void setReadAhead() throws IOException
+    {
+        mark(_currentBufferSize);
+        _readAhead = true;
+        try
+        {
+            _markedChar = _markedCharField.getInt(this);
+        }
+        catch (IllegalAccessException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void resetReadAhead() throws IOException
+    {
+        reset();
+        _readAhead = false;
+        _markedChar = -1;
+    }
+
+    @Override
+    public String readLine() throws IOException
```

Does this class need to override other read() methods?  Maybe they are safe, but first thought is that they should be overridden or throw Unsupported.

---

It's not meant to be a general purpose replacement for BufferedReader, so yes, probably should. skip() as well.